dnf-json

```diff
@@ -11,21 +11,83 @@ import logging
 import sys
 from http.server import BaseHTTPRequestHandler
 import pathlib
+from multiprocessing import Lock
+import queue
+import shutil
+from datetime import datetime, timedelta
 
 import dnf
 import hawkey
+import pickle
 
+# Logging setup (to systemd if available)
 formatter = logging.Formatter(
         fmt="%(asctime)s %(name)s.%(levelname)s: %(message)s",
         datefmt="%Y.%m.%d %H:%M:%S"
         )
 handler = logging.StreamHandler(stream=sys.stdout)
 handler.setFormatter(formatter)
-
 log = logging.getLogger('dnf-json')
 log.addHandler(handler)
 log.setLevel(logging.INFO)
 
+# Synchronisation necessary for the multiprocess request handling.
+process_lock = Lock()
+
+class CacheState():
+    """
+    A CacheState keeps track of the cache folders.
+    CacheState objects can update the list of used cache folders, which
+    will associate a date object to the name of the folder.
+    CacheState objects can ask to clean the unused cache folders.
+    Meaning that the folders having reach a certain timeout will be deleted.
+    """
+
+    def __init__(self, cache_dir, cache_timeout, folder_dict={}):
+        """
+        cache_dir the place where the state is stored
+        folder_dict a dict containing the existing list of cache folders
+        cache_timeout a timedelta before a cache folder can be deleted
+        """
+        self.cache_dir = cache_dir
+        self.folder_dict = folder_dict
+        self.cache_timeout = cache_timeout
+
+    def update_used(self, folder):
+        """
+        Associate a dateime.now() to the folders given as parameters
+        """
+        log.debug("Folder %s was used", folder)
+        self.folder_dict[folder] = datetime.now()
+
+    def clean_unused(self):
+        """
+        Delete the folders having reach the timeout
+        """
+        log.info("clean unused folders")
+        now = datetime.now()
+        list_folder_to_delete = []
+        for folder, then in self.folder_dict.items():
+            delta = now - then
+            log.debug("delete %s if %s > than %s", folder, delta, self.cache_timeout)
+            if delta > self.cache_timeout:
+                list_folder_to_delete.append(folder)
+        for folder in list_folder_to_delete:
+            del self.folder_dict[folder]
+            shutil.rmtree(folder)
+
+    @staticmethod
+    def load_cache_state_from_disk(cache_dir):
+        try:
+            with open(os.path.join(cache_dir,"cache_state.pkl"), "rb") as inp:
+                return pickle.load(inp)
+        except FileNotFoundError:
+            return CacheState(cache_dir, timedelta(hours=1))
```

Comment:
Isn't 1 hour too little? The cache isn't that big, I would go with 24 hours, or even more.

Reply:
It depends on the proportion of the onetimer repos in the requests we get. One hour will not delete repos that are frequently used anyway but will flush quickly the others. How often should we clean the cache now ?

Reply:
I bumped it to 24H :)